addon/components/polaris-dropzone.js

```diff
@@ -0,0 +1,522 @@
+
+    this.dragTargets = [];
+
+    this.get('state').setProperties({
+      dragging: false,
+      error: rejectedFiles.length > 0,
+    });
+
+    onDrop(files, acceptedFiles, rejectedFiles);
+
+    if (acceptedFiles.length) {
+      onDropAccepted(acceptedFiles);
+    }
+
+    if (rejectedFiles.length) {
+      onDropRejected(rejectedFiles);
+    }
+  },
+
+  handleDragEnter(event) {
+    event.preventDefault();
+    event.stopPropagation();
+
+    let { disabled, onDragEnter } = this.getProperties('disabled', 'onDragEnter');
+    if (disabled) {
+      return;
+    }
+
+    let fileList = getDataTransferFiles(event);
+    if (event.target && this.dragTargets.indexOf(event.target) === -1) {
+      this.dragTargets.push(event.target);
+    }
+
+    let state = this.get('state');
+    if (state.get('dragging')) {
+      return false;
+    }
+
+    let { rejectedFiles } = this.getValidatedFiles(fileList);
+
+    state.setProperties({
+      dragging: true,
+      error: rejectedFiles.length > 0,
+    });
+
+    onDragEnter();
+  },
+
+  handleDragOver(event) {
+    event.preventDefault();
+    event.stopPropagation();
+
+    let { disabled, onDragOver } = this.getProperties('disabled', 'onDragOver');
+    if (disabled) {
+      return;
+    }
+
+    onDragOver();
+
+    return false;
+  },
+
+  handleDragLeave(event) {
+    event.preventDefault();
+
+    let { disabled, onDragLeave, dropNode } = this.getProperties(
+      'disabled',
+      'onDragLeave',
+      'dropNode',
+    );
+    if (disabled) {
+      return;
+    }
+
+    this.dragTargets = this.dragTargets.filter((el) => {
+      return el !== event.target &&
+        dropNode &&
+        dropNode.contains(el);
+    });
+
+    if (this.dragTargets.length > 0) {
+      return;
+    }
+
+    this.get('state').setProperties({
+      dragging: false,
+      error: false,
+    });
+
+    onDragLeave();
+  },
+
+  handleDragStart(event) {
+    event.preventDefault();
+    event.stopPropagation();
+  },
+
+  adjustSize() {
+    throttle(this, function() {
+      let node = this.get('node');
+      let size = 'large';
+      let width = node.getBoundingClientRect().width;
+
+      if (width < 114) {
+        size = 'small';
+      } else if (width < 300) {
+        size = 'medium';
+      }
+
+      this.set('state.size', size);
+    }, 50);
+  },
+
+  getValidatedFiles(files) {
+    let { accept, allowMultiple, customValidator } = this.getProperties(
+      'accept',
+      'allowMultiple',
+      'customValidator',
+    );
+
+    let acceptedFiles = [];
+    let rejectedFiles = [];
+
+    Array.from(files).forEach((file) => {
+      if (!fileAccepted(file, accept) || (customValidator && !customValidator(file))) {
+        rejectedFiles.push(file);
+      } else {
+        acceptedFiles.push(file);
+      }
+    });
+
+    if (!allowMultiple) {
+      acceptedFiles.splice(1, acceptedFiles.length);
+      rejectedFiles.push(...acceptedFiles.slice(1));
```

Comment:
There _must_ be a more legible way of writing these... do you think something like

```js
let [ acceptedFile, ...leftoverFiles ] = acceptedFiles;
acceptedFiles = [ acceptedFile ];
rejectedFiles.push(...leftoverFiles);
```

makes it easier to follow what's going on? Otherwise comment these two lines please, they're not very clear IMO. Or otherwise, wrap this logic up in the loop above to just generate the accepted/rejected lists correctly to begin with instead of generating them then modifying them later?